name: tests+pypi

defaults:
  run:
    shell: bash

on:
  push:
    branches: [ main ]
  pull_request:
    branches: [ main ]
  schedule:
    - cron: '0 13 * * 4'
  release:
    types: [published]

jobs:
  build:
    strategy:
      matrix:
        mpi-np: [2, 3]
        platform:
          - ubuntu-latest
          - windows-latest
          - macos-13          # arch: Intel
          - macos-14          # arch: arm64
        python-version: ["3.9", "3.10", "3.11", "3.12", "3.13"]
        mpi: [ 'mpich', 'openmpi', 'msmpi', 'intelmpi']
        exclude:
        - platform: macos-13
          mpi: msmpi
        - platform: macos-13
          mpi: intelmpi
        - platform: macos-14
          mpi: msmpi
        - platform: macos-14
          mpi: intelmpi
        - platform: ubuntu-latest
          mpi: msmpi
        - platform: windows-latest
          mpi: mpich
        - platform: windows-latest
          mpi: openmpi
        - platform: windows-latest
          mpi: intelmpi

        # https://github.com/numba-mpi/numba-mpi/issues/69
        - platform: ubuntu-latest
          python-version: 3.9
          mpi: mpich
      fail-fast: false
    runs-on: ${{ matrix.platform }}
    steps:
      - uses: actions/checkout@v4
      - uses: actions/setup-python@v5
        with:
          python-version: ${{ matrix.python-version }}
      - uses: mpi4py/setup-mpi@v1
        with:
          mpi: ${{ matrix.mpi }}
      - if: matrix.mpi == 'intelmpi'
        run: |
          # checking if multiple MPI libs are OK
          sudo apt-get install mpi-default-bin
          mpiexec python -c "import os; assert 'I_MPI_ROOT' in os.environ"
      - run: pip install -e .[tests]
      - run: python -We -c "import mpi4py"
      - run: python -We -c "import numba_mpi"
      - run: mpiexec -n ${{ matrix.mpi-np }} pytest --durations=10 -p no:unraisableexception -We

  dist:
    runs-on: ubuntu-latest
    permissions:
      id-token: write
    steps:
      - uses: mpi4py/setup-mpi@v1
      - uses: actions/checkout@v2
        with:
          fetch-depth: 0  # https://github.com/pypa/setuptools_scm/issues/480
      - uses: actions/setup-python@v2
        with:
          python-version: "3.10"

      - run: pip install twine build

      - run: |
          unset CI
          python -m build 2>&1 | tee build.log
          exit `fgrep -i warning build.log | grep -v "tool.setuptools_scm" | wc -l`

      - run: twine check --strict dist/*

      - run: |
          python setup.py egg_info
          ex -sc 'g/^\[tests\]$/d' -cx numba_mpi.egg-info/requires.txt
          pip install -r numba_mpi.egg-info/requires.txt
          pip install --no-index --find-links file:./dist numba-mpi
          cd /tmp  # make sure local files are not picked up
          python -We -c "import numba_mpi"

      - if: github.event_name == 'push' && github.ref == 'refs/heads/main'
        uses: pypa/gh-action-pypi-publish@release/v1.12
        with:
          attestations: false
          repository_url: https://test.pypi.org/legacy/

      - if: startsWith(github.ref, 'refs/tags')
        uses: pypa/gh-action-pypi-publish@release/v1.12
        with:
          attestations: false
